### PR TITLE
21527 - Cancel/Apply EFT payment, initial for reverse payment

### PR DESCRIPTION
**Pull request by ochiu**
*Issue #:*
https://github.com/bcgov/entity/issues/21527

*Description of changes:*
- cleaned up services and moved EFT logic into EFTShortname Service.
- Cancel / Apply EFT payment actions
- Initial skeleton for reverse payment
- Added PATCH short name link for de-activating a link (removed old delete route)

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of the sbc-pay license (Apache 2.0).

### CODE REVIEW

**Comment by ochiu** [comment]
Will rebase against main

File: pay-api/src/pay_api/services/eft_service.py [hunk 1]
```diff
@@ -67,31 +67,6 @@ def create_invoice(self, payment_account: PaymentAccount, line_items: List[Payme
                        **kwargs) -> None:
         """Do nothing here, we create invoice references on the create CFS_INVOICES job."""
 
-    def apply_credit(self,
```
Comment (ochiu) [inline pay-api/src/pay_api/services/eft_service.py:70]: Not needed any more. Most of this is done in the jobs now.

Credit management moved to short name service as it makes more sense to live in there now.

File: pay-api/src/pay_api/services/eft_short_names.py [hunk 2]
```diff
@@ -68,6 +70,158 @@ class EFTShortnamesSearch:  # pylint: disable=too-many-instance-attributes
...
+        """Apply EFT payments to outstanding payments."""
+        current_app.logger.debug('<apply_payment_action')
+        if auth_account_id is None or PaymentAccountModel.find_by_auth_account_id(auth_account_id) is None:
+            raise BusinessException(Error.EFT_PAYMENT_ACTION_ACCOUNT_ID_REQUIRED)
+
+        cls.process_owing_statements(short_name_id, auth_account_id)
+        current_app.logger.debug('>apply_payment_action')
+
+    @classmethod
+    def _reverse_payment_action(cls, short_name_id: int, statement_id: int):  # pylint: disable=unused-argument
```
Comment (ochiu) [inline pay-api/src/pay_api/services/eft_short_names.py:208]: Will be implemented in upcoming ticket #21528.

File: pay-api/tests/unit/api/test_eft_short_names.py [hunk 3]
```diff
@@ -745,108 +812,3 @@ def test_search_eft_short_names(session, client, jwt, app):
                       data_dict['single-linked']['short_name'],
                       data_dict['single-linked']['accounts'][0],
                       data_dict['single-linked']['statement_summary'][0])
-
-
-@pytest.mark.skip(reason='This needs to be re-thought, the create cfs invoice job should be handling receipt creation'
-                         'and creating invoice references when payments are mapped, '
-                         'it should wait until 6 pm before marking invoices as PAID'
-                         'Otherwise calls to CFS could potentially fail and the two systems would go out of sync.')
-def test_apply_eft_short_name_credits(session, client, jwt, app):
```
Comment (ochiu) [inline pay-api/tests/unit/api/test_eft_short_names.py:754]: Not needed here. As stated in the comment a bunch of what this tests here no longer belongs here.
New tests in test_eft_payment_actions.py added.

File: pay-api/src/pay_api/resources/v1/eft_short_names.py [hunk 4]
```diff
@@ -189,3 +191,28 @@ def delete_eft_shortname_link(short_name_id: int, short_name_link_id: int):
...
+def post_eft_statement_payment(short_name_id: int):
+    """Perform EFT short name payment action on statement."""
+    current_app.logger.info('<post_eft_statement_payment')
+    request_json = request.get_json()
+
+    try:
+        if not EFTShortnameService.find_by_short_name_id(short_name_id):
+            response, status = {}, HTTPStatus.NOT_FOUND
+        else:
+            valid_format, errors = schema_utils.validate(request_json, 'eft_payment')
```
Comment (seeker25) [inline pay-api/src/pay_api/resources/v1/eft_short_names.py:208]: one day we should figure out how to streamline these 3 lines

File: pay-api/tests/unit/api/test_eft_payment_actions.py [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+
+def test_eft_apply_credits_action(db, session, client, jwt, app):
+    """Assert that EFT payment apply credits action works."""
+    token = jwt.create_jwt(get_claims(roles=[Role.MANAGE_EFT.value]), token_header)
+    headers = {'Authorization': f'Bearer {token}', 'content-type': 'application/json'}
+
+    account, short_name = setup_account_shortname_data()
+    setup_statement_data(account=account, invoice_totals=[50, 50, 100])
+
+    # Assert required accountId
```
Comment (seeker25) [inline pay-api/tests/unit/api/test_eft_payment_actions.py:90]: redundant

File: pay-api/tests/unit/api/test_eft_payment_actions.py [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+def test_eft_cancel_payment_action(session, client, jwt, app):
+    """Assert that EFT payment cancel action works."""
+    token = jwt.create_jwt(get_claims(roles=[Role.MANAGE_EFT.value]), token_header)
+    headers = {'Authorization': f'Bearer {token}', 'content-type': 'application/json'}
+
+    account, short_name = setup_account_shortname_data()
+    setup_statement_data(account=account, invoice_totals=[50, 50, 100])
+    eft_credits = setup_eft_credits(short_name=short_name, credit_amounts=[50, 25, 125])
+
+    # Assert required accountId
```
Comment (seeker25) [inline pay-api/tests/unit/api/test_eft_payment_actions.py:135]: redundant

File: pay-api/tests/unit/api/test_eft_payment_actions.py [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+    # Make EFT Payment
+    rv = client.post(f'/api/v1/eft-shortnames/{short_name.id}/payment',
+                     data=json.dumps({'action': EFTPaymentActions.APPLY_CREDITS.value,
+                                      'accountId': account.auth_account_id}),
+                     headers=headers)
+    assert rv.status_code == 204
+    assert all(eft_credit.remaining_amount == 0 for eft_credit in eft_credits)
+    assert EFTShortnamesService.get_eft_credit_balance(short_name.id) == 0
+
+    # Cancel Payment - Trigger unexpected credit amount
```
Comment (seeker25) [inline pay-api/tests/unit/api/test_eft_payment_actions.py:157]: redundant

File: pay-api/tests/unit/api/test_eft_payment_actions.py [hunk 8]
```diff
@@ -0,0 +1,225 @@
...
+    assert rv.status_code == 400
+    assert rv.json['type'] == Error.EFT_PAYMENT_ACTION_ACCOUNT_ID_REQUIRED.name
+
+    rv = client.post(f'/api/v1/eft-shortnames/{short_name.id}/payment',
+                     data=json.dumps({'action': EFTPaymentActions.CANCEL.value,
+                                      'accountId': account.auth_account_id}),
+                     headers=headers)
+    assert rv.status_code == 204
+
+    # Make EFT Payment
```
Comment (seeker25) [inline pay-api/tests/unit/api/test_eft_payment_actions.py:148]: redundant

File: pay-api/tests/unit/api/test_eft_payment_actions.py [hunk 9]
```diff
@@ -0,0 +1,225 @@
...
+
+    assert rv.status_code == 400
+    assert rv.json['type'] == Error.EFT_CREDIT_AMOUNT_UNEXPECTED.name
+    eft_credits[1].remaining_amount -= credit_offset
+    eft_credits[1].save()
+    # Assert no change and rollback was successful
+    assert all(eft_credit.remaining_amount == 0 for eft_credit in eft_credits)
+    assert EFTShortnamesService.get_eft_credit_balance(short_name.id) == 0
+
+    # Cancel EFT Payment
```
Comment (seeker25) [inline pay-api/tests/unit/api/test_eft_payment_actions.py:174]: redundant

**Review by seeker25 — COMMENTED** [review]
LGTM so far

**Review by seeker25 — APPROVED** [review]

**Pull request merged**